skimage/feature/sift.py

```diff
@@ -0,0 +1,730 @@
+                c_norm = np.subtract(c, center_pos[k, 1], dtype=float_dtype)
+                r_norm, c_norm = self._rotate(r_norm, c_norm, ori)
+
+                # select coordinates and gradient values within the patch
+                inside = np.maximum(np.abs(r_norm), np.abs(c_norm)) < rad_k
+                r_norm, c_norm = r_norm[inside], c_norm[inside]
+                r_idx, c_idx = np.nonzero(inside)
+                r = r[r_idx, 0]
+                c = c[0, c_idx]
+                gradient_row = gradient[0][r, c, scales[k]]
+                gradient_col = gradient[1][r, c, scales[k]]
+                # compute the (relative) gradient orientation
+                theta = np.arctan2(gradient_col, gradient_row) - ori
+                lam_sig = self.lambda_descr * sigma[k]
+                # Gaussian weighted kernel magnitude
+                kernel = np.exp((r_norm * r_norm + c_norm * c_norm)
+                                / (-2 * lam_sig ** 2))
+                magnitude = np.sqrt(gradient_row * gradient_row
+                                    + gradient_col * gradient_col) * kernel
+
+                lam_sig_ratio = 2 * lam_sig / self.n_hist
+                rc_bins = (hists - (1 + self.n_hist) / 2) * lam_sig_ratio
+                rc_bin_spacing = lam_sig_ratio
+                ori_bins = (2 * np.pi * bins) / self.n_ori
+
+                # distances to the histograms and bins
+                dist_r = np.abs(np.subtract.outer(rc_bins, r_norm))
+                dist_c = np.abs(np.subtract.outer(rc_bins, c_norm))
+
+                # the orientation histograms/bins that get the contribution
+                near_t, near_t_val = _ori_distances(ori_bins, theta)
+
+                # create the histogram
+                _update_histogram(histograms, near_t, near_t_val, magnitude,
+                                  dist_r, dist_c, rc_bin_spacing)
+
+                # convert the histograms to a 1d descriptor
+                histograms = histograms.flatten()
+                # saturate the descriptor
+                histograms = np.minimum(histograms,
+                                        0.2 * np.linalg.norm(histograms))
+                # normalize the descriptor
+                descriptor = (512 * histograms) / np.linalg.norm(histograms)
+                # quantize the descriptor
+                descriptor = np.minimum(np.floor(descriptor), 255)
+                self.descriptors[numbers[k], :] = descriptor
+
+    def detect(self, image):
+        """Detect the keypoints.
+
+        Parameters
+        ----------
+        image : 2D array
+            Input image.
+
+        """
+        check_nD(image, 2)
+        image = img_as_float(image)
+        float_dtype = _supported_float_type(image.dtype)
+        image = image.astype(float_dtype, copy=False)
+
+        self._set_number_of_octaves(image.shape)
+        self.deltas = self._deltas(float_dtype)
```

Comment:
These lines are repeated in `extract` and `detect_and_extract`. They can be refactored in a `_preprocess` function for example.

Reply:
Done!